staging/vhost-device-sound/src/device.rs

```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
+        const BACKEND: BackendType = BackendType::Null;
+
+        let config = SoundConfig::new(SOCKET_PATH.to_string(), false, BACKEND);
+
+        let chmaps = Arc::new(RwLock::new(vec![]));
+        let jacks = Arc::new(RwLock::new(vec![]));
+        let queue_indexes = vec![1, 2, 3];
+        let streams = Arc::new(RwLock::new(vec![Stream {
+            ..Stream::default()
+        }]));
+        let streams_no = 3;
+        let thread =
+            VhostUserSoundThread::new(chmaps, jacks, queue_indexes, streams.clone(), streams_no);
+
+        assert!(thread.is_ok());
+        let mut t = thread.unwrap();
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap(),
+        );
+        t.mem = Some(mem.clone());
+
+        let vrings = [
+            VringRwLock::new(mem.clone(), 0x1000).unwrap(),
+            VringRwLock::new(mem.clone(), 0x1000).unwrap(),
+            //VringRwLock::new(mem.clone(), 0x1000).unwrap(),
+        ];
+
+        let audio_backend =
+            RwLock::new(alloc_audio_backend(config.audio_backend, streams.clone()).unwrap());
+        assert!(t
+            .handle_event(CONTROL_QUEUE_IDX, &vrings, &audio_backend)
+            .is_ok());
+
+        let vring = VringRwLock::new(mem, 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+        assert!(t.process_control(&vring, &audio_backend).is_ok());
+        assert!(t.process_tx(&vring, &audio_backend).is_ok());
+        assert!(t.process_rx(&vring, &audio_backend).is_ok());
+    }
+    #[test]
+    fn test_sound_thread_failure() {
+        const SOCKET_PATH: &str = "vsoundfailure.socket";
+        const BACKEND: BackendType = BackendType::Null;
+
+        let config = SoundConfig::new(SOCKET_PATH.to_string(), false, BACKEND);
+
+        let chmaps = Arc::new(RwLock::new(vec![]));
+        let jacks = Arc::new(RwLock::new(vec![]));
+        let queue_indexes = vec![1, 2, 3];
+        let streams = Arc::new(RwLock::new(vec![]));
+        let streams_no = 0;
+        let thread =
+            VhostUserSoundThread::new(chmaps, jacks, queue_indexes, streams.clone(), streams_no);
+
+        let t = thread.unwrap();
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap(),
+        );
+
+        let audio_backend =
+            RwLock::new(alloc_audio_backend(config.audio_backend, streams.clone()).unwrap());
+
+        let vring = VringRwLock::new(mem, 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+        assert!(t.process_control(&vring, &audio_backend).is_err());
+        assert!(t.process_tx(&vring, &audio_backend).is_err());
+    }
+    #[test]
+    fn test_sound_backend() {
+        const SOCKET_PATH: &str = "vsound.socket";
+        const BACKEND: BackendType = BackendType::Null;
```

Comment:
And I think it would be better if there were no `const BACKEND` definitions and the value was inlined, like:

```rust
let config = SoundConfig::new(SOCKET_PATH.to_string(), 
false,
BackendType::Null);
```